src/engine/IdTable.h

```diff
@@ -0,0 +1,814 @@
+    }
+    IdTableImpl<COLS>::_manage_storage = true;
+  }
+
+  // Move constructor
+  IdTableStatic<COLS>(IdTableStatic<COLS>&& other) {
+    swap(*this, other);
+    other.IdTableImpl<COLS>::_data = nullptr;
+    other.IdTableImpl<COLS>::_size = 0;
+    other.IdTableImpl<COLS>::_capacity = 0;
+  }
+
+  // copy assignment
+  IdTableStatic<COLS>& operator=(IdTableStatic<COLS> other) {
+    swap(*this, other);
+    return *this;
+  }
+
+  // Element access
+  Id& operator()(size_t row, size_t col) {
+    return IdTableImpl<COLS>::_data[row * IdTableImpl<COLS>::_cols + col];
+  }
+  const Id& operator()(size_t row, size_t col) const {
+    return IdTableImpl<COLS>::_data[row * IdTableImpl<COLS>::_cols + col];
+  }
+
+  // Row access
+  typename IdTableImpl<COLS>::row_reference operator[](size_t row) {
+    return IdTableImpl<COLS>::getRow(row);
+  }
+
+  typename IdTableImpl<COLS>::const_row_reference operator[](size_t row) const {
+    // Moving this method to impl allows for efficient ConstRow types when
+    // using non dynamic IdTables.
+    return IdTableImpl<COLS>::getConstRow(row);
+  }
+
+  // Begin iterator
+  iterator begin() {
+    return iterator(IdTableImpl<COLS>::_data, 0, IdTableImpl<COLS>::_cols);
+  }
+  const_iterator begin() const {
+    return iterator(IdTableImpl<COLS>::_data, 0, IdTableImpl<COLS>::_cols);
+  }
+  const_iterator cbegin() const {
+    return iterator(IdTableImpl<COLS>::_data, 0, IdTableImpl<COLS>::_cols);
+  }
+
+  // End iterator
+  iterator end() {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+  const_iterator end() const {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+  const_iterator cend() const {
+    return iterator(IdTableImpl<COLS>::_data, IdTableImpl<COLS>::_size,
+                    IdTableImpl<COLS>::_cols);
+  }
+
+  typename IdTableImpl<COLS>::row_reference back() {
+    return IdTableImpl<COLS>::getRow((end() - 1).row());
+  }
+
+  typename IdTableImpl<COLS>::const_row_reference back() const {
+    return IdTableImpl<COLS>::getConstRow((end() - 1).row());
+  }
+
+  Id* data() { return IdTableImpl<COLS>::_data; }
+  const Id* data() const { return IdTableImpl<COLS>::_data; }
+
+  // Size access
+  size_t rows() const { return IdTableImpl<COLS>::_size; }
+
+  /**
+   * The template parameter here is used to allow for creating
+   * a version of the function that is constexpr for COLS != 0
+   * and one that is not constexpt for COLS == 0
+   */
+  using IdTableImpl<COLS>::cols;
+
+  size_t size() const { return IdTableImpl<COLS>::_size; }
+
+  /**
+   * @brief Sets the number of columns. Should only be called while data is
+   *        still nullptr.
+   **/
+  void setCols(size_t cols) {
+    assert(IdTableImpl<COLS>::_data == nullptr);
+    IdTableImpl<COLS>::setCols(cols);
+  }
+
+  void emplace_back() { push_back(); }
```

Comment:
Do we need the no param `emplace_back()` if we don't have one with params?

Reply:
I've added the no params version for convenience (and because it was used in several places in the code and made porting that to `IdTables` easier).  I haven't yet found a nice way of implementing an emplace with arguments that doesn't rely on c style va_args.